File: pdoo/ruby/specificdamage.rb
#encoding:utf-8

require_relative "damage"
require_relative "weapon"
require_relative "./lib/SpecificDamageToUI.rb"

module Deepspace

  class SpecificDamage<Damage

    public_class_method :new

    attr_reader :weapons
		
    def initialize(wl,s)
      super(s)
      @weapons = wl
    end
		
    def copy
      new SpecificDamage(@weapons,@nShields)
    end
		
    # Comprueba si el array 'w' contiene armas del tipo 't'
    private def arrayContainsType(w,t)
      w.each_index do |i|
        if w[i].type==t
          return i
        end
      end
      return -1
    end
    
    def adjust(w,s)
      new_shields = [@nShields,s.length].min
        new_weapons = Array.new(@weapons)

        new_weapons.each_with_index do |wt,i|
          if (arrayContainsType(w,wt)==-1)
            new_weapons[i] = nil
          else
            index = arrayContainsType(w,wt)
            w_aux = Array.new
            w.each_index do |i|
              if i>=index
                w_aux << w[i].type
              end
            end
            if (new_weapons.count(wt)>w_aux.count(wt))
              new_weapons[i] = nil
            end
          end
        end

        new_weapons = new_weapons.compact
        SpecificDamage.new(new_weapons,new_shields)
    end

    def discardWeapon(w)
      type = w.type
      index = -1
      @weapons.each_index do |i|
        if @weapons[i]==type
          index = i
        end
      end
      if index!=-1
        @weapons.delete_at(index)
      end
    end

    def hasNoEffect
      return (@weapons.length==0 and super)
    end

    def getUIversion
      SpecificDamageToUI.new(self)
    end
  end
end

File: pdoo/java/Deepspace/SpecificDamage.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package Deepspace;

import java.util.ArrayList;

/**
 * @brief Clase de daño específico
 */
public class SpecificDamage extends Damage {
    private ArrayList<WeaponType> weapons = new ArrayList<>();
    
    SpecificDamage(ArrayList<WeaponType> wl,int s) {
        super(s);
        weapons = wl;
    }
        
    private int arrayContainsType(ArrayList<Weapon> w,WeaponType t) {
        for (int i=0;i<w.size();i++) {
            if (w.get(i).getType()==t)
                return i;
        }
        return -1;
    }
    
    public ArrayList<WeaponType> getWeapons() { return weapons; }
    
    @Override
    public SpecificDamage copy() { return new SpecificDamage(weapons,getNShields()); }
 
    @Override
    public SpecificDamageToUI getUIversion() { return new SpecificDamageToUI(this); }
    
    @Override
    public SpecificDamage adjust(ArrayList<Weapon> w, ArrayList<ShieldBooster> s) {
        int newShields = (getNShields()<s.size())?getNShields():s.size();
        ArrayList<WeaponType> newWeapons = new ArrayList<>();
        
        for (int i=0;i<weapons.size();i++) {
            int index = arrayContainsType(w,weapons.get(i));
            if (index!=-1) {
                int tipoEnDanio = 0, tipoEnArmas = 1;
                for (WeaponType wt: newWeapons)
                    if (wt==weapons.get(i))
                        tipoEnDanio++;
                for (int j=index+1;j<w.size();j++)
                    if (w.get(j).getType()==weapons.get(i))
                        tipoEnArmas++;
                if (tipoEnDanio<tipoEnArmas)
                    newWeapons.add(weapons.get(i));
            }
        }
        
        return new SpecificDamage(newWeapons,newShields);
    }
    
    @Override
    public void discardWeapon(Weapon w) {
        boolean contained = false;
        
        for (int i=0;i<weapons.size() || !contained;i++)
            if (w.getType()==weapons.get(i)) {
                contained = true;
                weapons.remove(i);
            }
    }
    
    @Override
    public boolean hasNoEffect() { return (super.hasNoEffect() && weapons.isEmpty()); }
    
    @Override
    public String toString() {
        String aux = super.toString();
        aux += "\nArmas: ";
        for (WeaponType w: weapons) {
            aux += w.toString() + " ";
        }
        return aux;
    }
}